src/Sylius/Bundle/UserBundle/EventListener/DefaultUsernameORMListener.php

```diff
@@ -29,23 +32,36 @@ public function onFlush(OnFlushEventArgs $onFlushEventArgs)
         $entityManager = $onFlushEventArgs->getEntityManager();
         $unitOfWork = $entityManager->getUnitOfWork();
 
-        $entities = array_merge(
-            $unitOfWork->getScheduledEntityInsertions(),
-            $unitOfWork->getScheduledEntityUpdates()
-        );
+        $this->processEntities($unitOfWork->getScheduledEntityInsertions(), $entityManager, $unitOfWork);
+        $this->processEntities($unitOfWork->getScheduledEntityUpdates(), $entityManager, $unitOfWork);
+    }
+
+    /**
+     * @param array $entities
+     * @param EntityManagerInterface $entityManager
+     */
+    private function processEntities($entities, EntityManagerInterface $entityManager, UnitOfWork $unitOfWork)
+    {
+        foreach ($entities as $customer) {
+            if (!$customer instanceof CustomerInterface) {
+                continue;
+            }
 
-        foreach ($entities as $entity) {
-            if (!$entity instanceof CustomerInterface) {
+            $user = $customer->getUser();
+            if (null === $user) {
                 continue;
             }
 
-            $user = $entity->getUser();
-            if (null !== $user && $entity->getEmail() !== $user->getUsername()) {
-                $user->setUsername($entity->getEmail());
-                $entityManager->persist($user);
-                $userMetadata = $entityManager->getClassMetadata(get_class($user));
-                $unitOfWork->recomputeSingleEntityChangeSet($userMetadata, $user);
+            if ($customer->getEmail() === $user->getUsername() && $customer->getEmailCanonical() === $user->getEmailCanonical()) {
```

Comment:
`$customer->getEmailCanonical() === $user->getEmailCanonical()` will be always true, I assume you meant `$user->getUsernameCanonical()`